hbase-it/src/main/java/org/apache/hadoop/hbase/chaos/ChaosAgent.java

```diff
@@ -0,0 +1,596 @@
+    } catch (Shell.ExitCodeException e) {
+      String output = shell.getOutput();
+      throw new Shell.ExitCodeException(e.getExitCode(), "stderr: " + e.getMessage()
+        + ", stdout: " + output);
+    }
+    LOG.info("Executed Shell command, exit code: " + shell.getExitCode() +
+      " , output:" + shell.getOutput());
+
+    return new Pair<>(shell.getExitCode(), shell.getOutput());
+  }
+
+  private <E extends Exception> void retryOrThrow(RetryCounter retryCounter, E ex,
+    String user, String cmd) throws E {
+    if (retryCounter.shouldRetry()) {
+      LOG.warn("Local command: " + cmd + " , user:" + user
+        + " failed at attempt " + retryCounter.getAttemptTimes() + ". Retrying until maxAttempts: "
+        + retryCounter.getMaxAttempts() + ". Exception: " + ex.getMessage());
+      return;
+    }
+    throw ex;
+  }
+
+  private boolean isConnected() {
+    return connected;
+  }
+
+  @Override
+  public void close() throws IOException {
+    LOG.info("Closing ZooKeeper Connection for Chaos Agent : " + agentName);
+    try {
+      zk.close();
+    } catch (InterruptedException e) {
+      LOG.error("Error while closing ZooKeeper Connection.");
+    }
+  }
+
+  @Override
+  public void run() {
+    try {
+      LOG.info("Running Chaos Agent on : " + agentName);
+      while (!this.isConnected()) {
+        Thread.sleep(100);
+      }
+      this.getTasks();
+      while (!stopChaosAgent.get()) {
+        Thread.sleep(500);
+      }
+    } catch (InterruptedException e) {
+      LOG.error("Error while running Chaos Agent");
+    }
+
+  }
+
+  @Override
+  public void process(WatchedEvent watchedEvent) {
+    LOG.info("Processing event: " + watchedEvent.toString());
+    if (watchedEvent.getType() == Event.EventType.None) {
+      switch (watchedEvent.getState()) {
+        case SyncConnected:
+          connected = true;
+          break;
+        case Disconnected:
+          connected = false;
+          break;
+        case Expired:
+          connected = false;
+          LOG.error("Session expired creating again");
+          try {
+            createZKConnection(null);
+          } catch (IOException e) {
+            LOG.error("Error creating Zookeeper connection");
+          }
+        default:
+          LOG.error("Unknown State");
+          break;
+      }
+    }
+  }
+
+  private void recreateZKConnection() throws Exception{
+    try {
+      zk.close();
+    } catch (InterruptedException e) {
+      LOG.error("Error closing ZK connection : " + e);
+      throw new RuntimeException(e) ;
+    } finally {
+      try {
+        createZKConnection(newTaskCreatedWatcher);
+        createEphemeralZNode(ChaosConstants.CHAOS_AGENT_REGISTRATION_EPIMERAL_ZNODE +
+          ChaosConstants.ZNODE_PATH_SEPARATOR + agentName, new byte[0]);
+      } catch (IOException e) {
+        LOG.error("Error creating new ZK COnnection for agent: " + agentName + e);
+        throw new RuntimeException(e);
+      }
+    }
+  }
```

Comment:
Throwing RTE within `finally` might complicate debugging. Let's keep everything in `try` block:
```
  private void recreateZKConnection() throws Exception {
    try {
      zk.close();
      createZKConnection(newTaskCreatedWatcher);
      createEphemeralZNode(ChaosConstants.CHAOS_AGENT_REGISTRATION_EPIMERAL_ZNODE +
        ChaosConstants.ZNODE_PATH_SEPARATOR + agentName, new byte[0]);
    } catch (InterruptedException | IOException e) {
      LOG.error("Error recreating new ZK Connection for agent: {}", agentName, e);
      throw e;
    }
  }
```